icicle/include/fields/field_config.cuh

```diff
@@ -0,0 +1,58 @@
+#pragma once
+#ifndef FIELD_CONFIG_H
+#define FIELD_CONFIG_H
+
+#include "fields/id.h"
+#include "gpu-utils/sharedmem.cuh"
+
+#include "fields/field.cuh"
+#if defined(EXT_FIELD)
+#include "fields/extension_field.cuh"
+#endif
+
+#if FIELD_ID == BN254_FIELDS
+#include "fields/params/bn254_fields.cuh"
+using namespace bn254;
+#elif FIELD_ID == BLS12_381_FIELDS
+#include "fields/params/bls12_381_fields.cuh"
+using namespace bls12_381;
+#elif FIELD_ID == BLS12_377_FIELDS
+#include "fields/params/bls12_377_fields.cuh"
+using namespace bls12_377;
+#elif FIELD_ID == BW6_761_FIELDS
+#include "fields/params/bw6_761_fields.cuh"
+using namespace bw6_761;
+#elif FIELD_ID == GRUMPKIN_FIELDS
+#include "fields/params/grumpkin_fields.cuh"
+using namespace grumpkin;
+#endif
+
+/**
+ * @namespace field_config
+ * Namespace with type definitions for finite fields. Here, concrete types are created in accordance
+ * with the `-DFIELD` env variable passed during build.
+ */
+namespace field_config {
+  /**
+   * Scalar field. Is always a prime field.
+   */
+  typedef Field<fp_config> scalar_t;
+
+#if defined(BASE_FIELD)
+  /**
+   * Base field of G1 curve. Is always a prime field.
```

Comment:
I find it weird that the field knows about the curve. I mean fields and curves are still somewhat coupled.
On the other hand we kind of discussed it and we go back and forth. I don't know  best option would be.

BTW is there one interesting extension field per field? I know that for some small fields the extension has 4 elements and wonder how it will fit here, and whether we will need more than one extension for a given field.

Reply:
I also find this curve/field binding very awkward.

Regarding the extension field, yes, I also thought that it should be more flexible. But I think it can be added afterwards, without breaking the structure

The curves' fields are not used on their own, that is true. Actually, that was my very first proposal - to not move their compilation into fields module, but rather compile them inside the according curve. But then, we started going back and forth with separating them into their own lib to utilize cmake caching (for grumpkin and bw).

Actually, I think I will once again try to move them into curves and only leave goldilocks/babybear/etc in `fields/CMakeLists.txt`

Reply:
I also think it's probably better but maybe let's discuss it with the team so that you don't do this change back and forth

Reply:
ok, I'll fix the other stuff then, while we wait for tomorrow's meeting